MAINT: backports for SciPy 1.16.1

tylerjereddy opened the pull request:
The time for SciPy `1.16.1` is probably drawing closer given the recent release of ABI-stable CPython `3.14.0rc1`, and also just the general accumulation of bugs. For the pre-requisite NumPy release that supports `3.14`, Chuck mentioned in the NumPy meeting this morning a few things about `cibuildwheel` (lack of stable release maybe?), and some Cirrus->GHA migration stuff happening there. Anyway, I'll keep an eye on the supporting NumPy release progress.

Backports included (so far, and with merge conflicts noted):

1. gh-23206
2. gh-23279
3. gh-23284
4. gh-23322
5. gh-23332 (**merge conflicts resolved manually**)
6. gh-23339
7. gh-23348
8. gh-23167
9. gh-23266
10. gh-23276
11. gh-23280
12. gh-23293
13. gh-23387
14. gh-23384
15. gh-23380
16. gh-23187

Backports already merged to the maintenance branch:
1. gh-23222

TODO:

- [x] backports needed for building wheels/supporting `3.14`? may still need a bit of action on `main` as well?
- [x] any other unmerged backports? I'd like to see gh-23293 backported, that fixes a pretty egregious bug, but it is my PR so would prefer not to self-merge
- [x] free-threading status/concerns? We ended up disabling the `3.13t` CI tests on the release branch here to just get things released--do we want to try to re-enable (it has been re-enabled on `main`, though I think just quite recently with Nathan's push) and also support `3.14t`? This may be quite a mess of backports?
- [x] update the release notes when the backporting work is done, and to add a short blurb about `3.14` support
- [x] get the CI passing, up to and including confirmation of `3.14`/`3.14t` wheel build/test success I suppose--that's going to be a decent number of wheels with `3.13`/ `3.13t` as well?

andyfaff committed:
BUG: shgo Complex cache split_edge differently

rgommers committed:
CI: enable wheel builds for Python 3.14-dev

rgommers committed:
TST: stats: skip one sampling.FastGeneratorInversion` test on musllinux

rgommers committed:
TST: io.matlab: skip zlib overlap test on Windows + Python 3.14

rgommers committed:
CI: exclude musllinux + cp314t wheel builds [wheel build]

Follow up on this later, likely one or more issues in CPython, NumPy,
and/or Pythran.

The aarch64 job was hanging at various places in the test suite.
The x86-64 job was failing on `pythran` generation .cpp code at build
time, repeatedly and on multiple source files.

mdhaber committed:
DOC: ndimage.vectorized_filter: correct `output` description

[docs only]

jorenham committed:
BUG: fix signature of ``@_transition_to_rng`` functions

ilayn committed:
BUG: linalg: Fix pointer casting order for sqrtm

dschmitz89 committed:
BUG: stats.beta.entropy: vectorize implementation (#23280)

Vectorize implementation to enable use of `make_distribution`

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>land <mhaberla@calpoly.edu>

andyfaff committed:
DOC: minor nit in least_squares

gmgunter committed:
BUG: signal.remez: fix handling of `weight` array (#23266)

* BUG: signal.remez: fix handling of `weight` array

Fix an issue in `signal.remez` that caused a spurious exception to be
raised if the `weight` argument was a NumPy array with more than one
element.

* TST: add regression test for gh-23266

* TST: test with various array API backends

dschult committed:
add test of format of multiply result, and fix the resulting failure. Fixes #23321

Greenie0701 committed:
BLD: Add bigobj flag for MSVC to fix object file section limit

robfalck committed:
Update minimize docstring

The docstring for minimize has been updated to reflect the fact that COBYLA now allows equality constraints

dschult committed:
Shift from using format to attribute presence and try/except (#23348)

rgommers committed:
CI: stop skipping cp314/cp314t musllinux-aarch64 wheels [wheel build]

rgommers commented:
Thanks Tyler. What's included as backports so far LGTM. I'm not sure what is up with the Oldest GCC job failure:
```
ERROR Missing dependencies:
	pybind11<2.14.0,>=2.13.2
```
However it's a good reminder that all upper bounds should be bumped. For `pybind11` we want to allow 3.0.0.



> backports needed for building wheels/supporting `3.14`? may still need a bit of action on `main` as well?

I think so, I'll have a look at that. And will look at the `ndimage` bug fix PR as well.

> free-threading status/concerns? 

No major concerns. I will check for fixes to backport; I don't want to re-enable the `pytest-run-parallel` job in a maintenance branch, it's not necessary and it won't be reliable (we don't want to backport the larger changes that enabled it in `main`).

> that's going to be a decent number of wheels with `3.13`/ `3.13t` as well?

Yes, there are going to be a lot, that is what it is. We should have enough space on PyPI still for this release, and the limit increase I submitted a month ago is approaching the front of the queue.

rgommers commented:
Two of the `cp314t-musllinux` jobs are still failing when I re-enable them, one with a Pythran-related build error and one with a hang in the test suite. So unfortunately there are some unfixed bugs left upstream: [CI log](https://github.com/rgommers/scipy/actions/runs/16449819858/job/46491759180).

rgommers committed:
CI: bump cibuildwheel to v3.1.0, for Python 3.14.0rc1

rgommers committed:
CI: disable special-casing of 3.14-dev now that Python 3.14rc1 is out

This reverts changes made in scipy#23187, keeping the commented out code
for once 3.15-dev arrives.

[ci skip]

tylerjereddy committed:
MAINT: Python 3.14 to classifiers

* With NumPy `2.3.2` now out with Python `3.14.0rc1` support,
and Ralf's PRs that test SciPy wheels against that same
version of Python now passing/merged, it is probably time
to add this to the classifiers and backport it.

* See: https://github.com/numpy/numpy/pull/29444

[ci skip] [skip ci]

tylerjereddy committed:
BUG: ndimage.median_filter: single element array handling (#23293)

Fixes gh-23075

This patch adds special handling for the single-element array scenario
in the above ticket, along with several regression test cases that get
their expected values from running the same tests against SciPy
`1.14.1`, before we overhauled the `median_filter` algorithm.

tylerjereddy committed:
MAINT, DOC: PR 23378 revisions

* Removed an unused import that snuck through
during backporting activity.

* Bump the allowed upper version bound for pybind11,
since a new release is out, and appears to be compatible
with the release branch of SciPy here.

* Update the SciPy 1.16.1 release notes following more
backport activity.

tylerjereddy committed:
DOC: PR 23378 revisions

* Mention support for CPython 3.14.0rc1
in the release notes for SciPy 1.16.1.

tylerjereddy commented:
Ok, this PR is now at parity with all merged backport PRs, and the release notes have been updated to reflect all of that backport activity + a brief mention of `3.14.0rc1` support.

Local testing on ARM Mac with Python `3.14.0rc1` seems "ok" on this branch. I think other comments have been address too--`pybind11` is the only lib I found where we needed to bump the upper version bound we allow because it had a recent new release.

If the regular CI is green, then I'll proceed with a test wheel build later, which is important with the addition of `3.14` to the matrix on the release branch.

tylerjereddy committed:
MAINT: PR 23378 wheel build test [wheel build]

[wheel build]

tylerjereddy commented:
Regular CI was all green, proceeding with full wheel build test.

tylerjereddy commented:
Wheel builds initiated for `1.16.1`: https://github.com/scipy/scipy/actions/runs/16546171812

I'll resume the release process in the morning.
